libraries/FSTools/FSTools.cpp

```diff
@@ -0,0 +1,253 @@
+    FSTOOLSDEBUG("Size of dummy FS = %u\n", tempFSsize);
+
+    FS tempFS = FS(FSImplPtr(new littlefs_impl::LittleFSImpl(startSector, tempFSsize, FS_PHYS_PAGE, FS_PHYS_BLOCK, 5)));
+
+    if (tempFS.format() && tempFS.begin())
+    {
+        if (_copyFS(*_pFS, tempFS))
+        {
+            FSTOOLSDEBUG("Files copied to temp File System\n");
+            reset();
+            if (destinationFS.format() && destinationFS.begin()) //  must format then mount the new FS
+            {
+                if (_copyFS(tempFS, destinationFS))
+                {
+                    FSTOOLSDEBUG("Files copied back to new FS\n");
+                    result = true;
+                }
+            }
+            else
+            {
+                FSTOOLSDEBUG("Error Mounting\n");
+            }
+        }
+        else
+        {
+            FSTOOLSDEBUG("Copy Failed\n");
+        }
+        tempFS.end();
+    }
+    else
+    {
+        FSTOOLSDEBUG("Failed to begin() TempFS\n");
+    }
+    return result;
+};
+
+void FSTools::reset()
+{
+    _mounted = false;
+    _layout = nullptr;
+    if (_pFS)
+    {
+        _pFS->end();
+        _pFS.reset();
+    }
+}
+
+void FSTools::fileListIterator(FS & fs, const char * dirName, FST::FileCb Cb)
+{
+    Dir dir = fs.openDir(dirName);
+    while (dir.next())
+    {
+        if (dir.isFile())
+        {
+            File f =  dir.openFile("r");
+            if (Cb)
+            {
+                Cb(f);
+            }
+        }
+        else
+        {
+            fileListIterator(fs, dir.fileName().c_str(), Cb);
+        }
+    }
+}
+
+
+uint32_t FSTools::_getStartAddr(const FST::layout & layout)
+{
+    return (layout.startAddr - 0x40200000);
+}
+
+uint32_t FSTools::_getSize(const FST::layout & layout)
+{
+    return (layout.endAddr - layout.startAddr);
+}
+
+#ifdef DEBUG_ESP_CORE
+void FSTools::_dumpFileInfo(File & f)
+{
+    if (f)
+    {
+        DEBUG_ESP_PORT.printf_P(PSTR(" File: %-30s [%8uB]\n"), f.fullName(), f.size());
+    }
+}
+#endif
+
+bool FSTools::_copyFS(FS & sourceFS, FS & destFS)
+{
+    uint32_t sourceFileCount = 0;
+    uint32_t sourceByteTotal = 0;
+
+    fileListIterator(sourceFS, "/", [&sourceFileCount, &sourceByteTotal](File & f)
+    {
+        if (f)
+        {
+            sourceFileCount++;
+            sourceByteTotal += f.size();
+        }
+    });
+
+    size_t count = 0;
+    fileListIterator(sourceFS, "/", [&count, &destFS](File & sourceFile)
+    {
+        if (sourceFile)
+        {
+            File destFile = destFS.open(sourceFile.fullName(), "w");
+            if (destFile)
+            {
+                destFile.setTimeout(5000); //  this value was chosen empirically as it failed with default timeout. 
+                size_t written = destFile.write(sourceFile);
+                if (written == sourceFile.size())
+                {
+                    count++;
+                }
+            }
```

Comment:
On a larger filesystem, or one with many files, it seems like this loop over the iterator could take a significant amount of time.  Adding a `yield()` after each file here might be prudent.  Have you done 200-400k filesystem migrations to see if this sometimes happens, or if I'm just being paranoid?

Reply:
This is a good point.  My tests do take some time with ~250k files, and even with WiFi active, maintaining a TCP connection and doing it from a loop it has been ok.  But I think it is prudent to add a yield tbh.